src/common/Language/Java/Inline/Internal.hs

```diff
@@ -139,12 +140,18 @@ loadJavaWrappers = doit `seq` return ()
   where
     {-# NOINLINE doit #-}
     doit = unsafePerformIO $ push $ do
+      idx <- newIORef 0
+      Magic.forEachDotClass $ const $ modifyIORef' idx (+1)
+      classRefs :: JObjectArray <- readIORef idx >>= newArray >>= newGlobalRefNonFinalized
+      writeIORef idx 0
       loader :: J ('Class "java.lang.ClassLoader") <- do
         thr <- callStatic "java.lang.Thread" "currentThread"
         call (thr :: J ('Class "java.lang.Thread")) "getContextClassLoader"
       Magic.forEachDotClass $ \Magic.DotClass{..} -> do
-        _ <- defineClass (referenceTypeName (SClass className)) loader classBytecode
-        return ()
+        cls <- defineClass (referenceTypeName (SClass className)) loader classBytecode
+        i <- readIORef idx
+        setObjectArrayElement classRefs i cls
+        modifyIORef idx (+1)
```

Comment:
`forEachDotClass` is only used in this spot. It could be changed to make this simpler.

Maybe
```Haskell
classRefs <- newArray (length Magic.dotClasses) >>= newGlobalRefNonFinalized
forM_ (zip Magic.dotClasses [0..]) $ \Magic.DotClass{..} i -> do
  cls <- defineClass (referenceTypeName (SClass className)) loader classBytecode
  setObjectArrayElement classRefs i cls
```

Also, note that the local references created with `defineClass` and `newArray` are still not deleted.

Reply:
> Also, note that the local references created with defineClass and newArray are still not deleted.

Actually, I'm just noticing the calls to push and pop in this code, so local references are being deleted in the end. It still would be preferable to delete each reference after writing it to the array, instead of accumulating all the references until the end.